packages/jest-runtime/src/index.js

```diff
@@ -673,18 +673,38 @@ class Runtime {
   }
 
   _createRequireImplementation(
-    from: Path,
+    from: Module,
     options: ?InternalModuleOptions,
   ): LocalModuleRequire {
     const moduleRequire =
       options && options.isInternalModule
-        ? (moduleName: string) => this.requireInternalModule(from, moduleName)
-        : this.requireModuleOrMock.bind(this, from);
+        ? (moduleName: string) =>
+            this.requireInternalModule(from.filename, moduleName)
+        : this.requireModuleOrMock.bind(this, from.filename);
     moduleRequire.cache = Object.create(null);
     moduleRequire.extensions = Object.create(null);
-    moduleRequire.requireActual = this.requireModule.bind(this, from);
-    moduleRequire.requireMock = this.requireMock.bind(this, from);
-    moduleRequire.resolve = moduleName => this._resolveModule(from, moduleName);
+    moduleRequire.requireActual = this.requireModule.bind(this, from.filename);
+    moduleRequire.requireMock = this.requireMock.bind(this, from.filename);
+    moduleRequire.resolve = moduleName =>
+      this._resolveModule(from.filename, moduleName);
+    Object.defineProperty(
```

Comment:
Instead of a traversal, couldn't we add the main module as an instance variable to the runtime instance and then just return it directly?

Reply:
This would require a to pass some flag in 3rd argument in other packages as well: https://github.com/facebook/jest/blob/master/packages/jest-circus/src/legacy_code_todo_rewrite/jest_adapter.js#L70 and https://github.com/facebook/jest/blob/master/packages/jest-jasmine2/src/index.js#L154
Since we cannot just assume that the first required module is main due to setup files.

Reply:
Got it, yeah I can live with this then.